### PR TITLE
CRM-20036 Don't retrieve all soft contributions if not needed.

### PR DESCRIPTION
**Pull request by johanv**
This is a fix for CRM-20036. See JIRA for details.

---

 * [CRM-20036: Memory problem when updating a contribution without soft contributions](https://issues.civicrm.org/jira/browse/CRM-20036)

### CODE REVIEW

File: CRM/Contribute/BAO/ContributionSoft.php [hunk 1]
```diff
@@ -236,11 +236,14 @@ public static function retrieve(&$params, &$defaults) {
...
     $contributionSoft = new CRM_Contribute_DAO_ContributionSoft();
-    $contributionSoft->id = CRM_Utils_Array::value('id', $params);
-    if (!$contributionSoft->find()) {
-      return FALSE;
+    if (!empty($params['id'])) {
+      $contributionSoft->id = CRM_Utils_Array::value('id', $params);
+      if (!$contributionSoft->find()) {
+        return FALSE;
+      }
+      // Are we really sure about this??
```
Comment (JoeMurray) [inline CRM/Contribute/BAO/ContributionSoft.php]: Good question. My guess is that this is to handle case where the soft contributions are successfully deleted and we don't want the url to cause an error by referring to something just deleted. However, the deletion might fail, in which case the url should go back to same soft contribution. Could you move the unset to between L250-251 ?

**Review by JoeMurray — COMMENTED** [review]
This looks like a good fix from a code review perspective, assuming the unset is dealt with.

**Comment by eileenmcnaughton** [comment]
What about this as an alternative? 

https://github.com/eileenmcnaughton/civicrm-core/commit/f139ac07dfc54cc1b7f9ff12e4bc490f1e4593ce

Basically that del function is called in every case except the one that I assume is causing this problem with the 'id' field set, so it seems to be adding code complexity without value. If we just use the api we can remove the function entirely

**Comment by eileenmcnaughton** [comment]
Opened #10260 as an alternative fix - closing this

**Pull request closed without merging**